distribution/xfer/download.go

```diff
@@ -62,6 +63,9 @@ type DownloadDescriptor interface {
 	// if it is unknown (for example, if it has not been downloaded
 	// before).
 	DiffID() (layer.DiffID, error)
+	// ForeignSource returns nil or the foreign layer's information and
+	// source.
+	ForeignSource() *distribution.Descriptor
```

Comment:
Do we want to bother storing a descriptor in non-foreign cases? That might be tricky since we don't know what the digest is going to be until we actually construct and compress the layer tar. It seemed simpler to only store the descriptor in the foreign case.